[FIX] mass_mailing: fix mail template scroll

amdi-odoo opened the pull request:
Purpose
=======
Fix the height of the mail templates in the mailing form.

Specification
=============
For small templates / theme selector, there is a blank space below the editor
because the iframe's height is matching the full height of the website snippets on the side.
For big templates, the website snippets doesn't follow the template scrolling which
makes it difficult to drag and drop the snippets.

This happens following the removal of the full width view and a related js file which
was previously resizing the editor height accordingly (ref commit: https://github.com/odoo/odoo/commit/cf6cd5ef46287f4e71fe96ad6b404ba9b08c1066 )

Reverting a part of the old "mailing_mailing_view_form_full_width" file into
the "mass_mailing_html_field" one so that the editor iframe is correctly resized
depending on its content and the editor sidebar is correctly resized to follow
the scrolling behavior.
Concretely, introducting back the resize observer and the "_updateIframe", "_isFullScreen",
"_resizeMailingEditorIframe" and "_repositionMailingEditorSidebar" methods.

Task-4086341

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/170901.png)](https://mergebot.odoo.com/odoo/odoo/pull/170901)

reth-odoo commented on addons/mass_mailing/static/src/js/mass_mailing_html_field.js:
```diff
@@ -25,15 +25,20 @@ export class MassMailingHtmlField extends HtmlField {
+            onIframeUpdated: this._updateIframe.bind(this),
             onWysiwygReset: this._resetIframe.bind(this),
         });
+        super.setup();
```
Is this really necessary ? Seems like fullwidth version worked fine with the super call at the top

amdi-odoo replied:
Hello @reth-odoo :slightly_smiling_face: 
The issue I had here was that in the super call, the `onIframeUpdated` was retrieved from the env and assigned:
https://github.com/odoo/odoo/blob/saas-17.2/addons/web_editor/static/src/js/backend/html_field.js#L70 
But at that time the `useSubEnv` hasn't been called yet so it was setting it to an empty method instead of using `_updateIframe` :confused: 
I didn't know how to face the issue so I just declared the setup method after the `useSubEnv` :sweat_smile: 

But now that I think about it, it'll maybe work to just overwrite the `this.onIframeUpdated` :thinking:
I'm trying that! :smile:

amdi-odoo replied:
@reth-odoo hi again :slightly_smiling_face: 
Well... Someone broke the display in 17.2 https://tinyurl.com/2cv3tyny 
Investigating why as it broke my fix as well :sweat_smile: (2 vertical scroll bars and one horizontal instead of only the one from the website snippets)
![image](https://github.com/odoo/odoo/assets/107679666/fb9a9557-deb2-467c-8dbc-f96b225cd15d)

reth-odoo replied:
Just realized it was a different controller for the full-width view, so it's not the same

Indeed here as we don't need to pass the value to children components just overwriting should be fine :crossed_fingers:

reth-odoo reviewed (commented):
I seem to remember there was a distinction because it looked weird in marketing_automation but it doesn't seem to really affect it so seems fine

amdi-odoo commented:
Hello @reth-odoo :slightly_smiling_face: 

The fix I mentioned [here](https://github.com/odoo/odoo/pull/170901#discussion_r1673965629) that broke the display has been reverted as it introduced too many issues https://github.com/odoo/odoo/pull/173109/commits/298fa4e3925e7f7fc20816594cfe0e7bfb298529 
I added in this PR a new commit to give another try to the previous fix.

For the first commit, I updated to overwrite the `this.onIframeUpdated` as mentioned in a previous comment :smile: 

Thanks and cheers!

amdi-odoo commented:
NB: For the commit regarding the compressed mode, in master, it would maybe be best to change all the `lg` bootstrap rules in the themes templates by `md` ones to best fix the new iframe width.

reth-odoo commented on addons/mass_mailing/static/src/js/mass_mailing_html_field.js:
```diff
@@ -34,6 +34,12 @@ export class MassMailingHtmlField extends HtmlField {
         this.orm = useService('orm');
         this.dialog = useService('dialog');
 
+        this.onIframeUpdated = this._updateIframe.bind(this);
```
I guess to be "correct" we might want to call the original function too, if possible

reth-odoo reviewed (commented):
Alright, style change seems fine thanks for looking into all of that ^^

Just a last nitpick, just in case

amdi-odoo commented:
Hello @awa-odoo :slightly_smiling_face: 

FYI I found an alternative to fix the compressed mode display https://github.com/odoo/odoo/pull/170901/commits/db37cb8c7a8fc8cd8d4ab7e1ea9581f25b404b05 

NB: still considering changing all the `lg` to `md` rules in master though as this is not a perfect fix (if you reduce a bit the window width it's going back to mobile view even though it could keep the large one)

Cheers! :smile:

tde-banana-odoo commented:
@abd-msyukyu-odoo You are our only hope here (at least to review).

abd-msyukyu-odoo commented:
@tde-banana-odoo Heya! Ok, I'll check this :)

I was planning on talking about it during the appraisal, but I don't think I'm in the correct github group to receive notifications about the new modules affected to the sm-engagement squad

tde-banana-odoo commented:
@abd-msyukyu-odoo Indeed it is not updated :D .

abd-msyukyu-odoo commented:
@amdi-odoo Hello, I tried a bunch of stuff, and honestly, the simplest solution I found which results in a usable view is to set the height of the snippet menu to 70vh in non-mobile, non-fullscreen mode.
This will force the iframe to also always be 70vh, which seems reasonable enough to work in.
The advantage of doing it like this is that we don't have to add a bunch of calculations which are performance intensive.
Granted, it will not look the same as in 17.0, but since the snippets menu was removed from the iframe, we have to "fight" against the default styles of the form view coming from above, and it either requires a lot of style overrides on the ancestors, either ugly JS computations, and both are not ideal.

in fullscreen mode, I found that the iframe was not centered in the space left by the snippets menu, if you set the `d-flex` parent of the iframe with `position-absolute`, `width: 100%` and `height: 100%`, and then remove the `width: 100%` of the iframe, it gives a better result.

In mobile, I don't know if we are supposed to see the snippets menu ? It seems there, but without dimensions.

About the chatter new dimensions, seems ok to me.

cc @tde-banana-odoo @awa-odoo 
Do you think it's worth introducing a ResizeObserver and compute things in javascript to mimick how the elements were placed in 17.0, or is it ok to change things a bit without loosing usability by using the 70vh strat ?

I guess it's a question for Luc, which isn't there ;p

Cheers!

reth-odoo commented:
> In mobile, I don't know if we are supposed to see the snippets menu ? It seems there, but without dimensions.

In mobile mode you're supposed to get the toolbar without any snippet. Idea is the snippet menu takes up too much space and you wouldn't actually ever want to edit a mailing on your phone beyond updating text.

abd-msyukyu-odoo commented:
Also, for context, why is it so complicated to style this view:
- the form view is designed to grow with the size of its content, and start scrolling on overflows
- we would like the snippet menu to always use the maximum available space in the viewport, without going outside, but it is placed in a parent which grows with its content (contradictory, the parent should use a set amount of space and the snippets menu should take 100% of it, not the reverse)
- the iframe is also a content which is designed to grow with the size of the parent, not the reverse, but is placed in a parent which grows with the size of its content)

awa-odoo commented:
@abd-msyukyu-odoo Thanks for the detailed analysis, even though I'm not sure to understand all the implications ^^

As things might (will?) change when implementing the new editor, would say we go for the easiest and shortest solution, which seems to be the 70vh setting.

amdi-odoo commented:
Hello @abd-msyukyu-odoo and @awa-odoo :slightly_smiling_face: 

I changed my implementation with the 70vh solution as it indeed implies less diff :smile: 

At first I only set the 70vh on the website snippets as suggested, but I spotted that it doesn't fix the white gap under the theme selector iframe because this one doesn't have the snippets on the side yet.
<img src="https://github.com/user-attachments/assets/13572d2b-47a4-4f85-92f0-0146c1d142ee" width="500" height="350">
As setting the 70vh on the snippets menu is forcing the iframe to be 70vh anyway, I just directly set an height of 70vh on the iframe as well so that the theme selector benefits from it :slightly_smiling_face: (not on mobile and fullscreen)

I tried your proposition with the `position: absolute` and the `width: 100%`, `height: 100%` but while doing so I spotted a [rule](https://github.com/odoo/odoo/blob/saas-17.2/addons/web_editor/static/src/scss/web_editor.backend.scss#L71) in web_editor directly targeting the mass mailing fullscreen iframe with the correct width :slightly_smiling_face: It was overwritten by the 100% width [rule](https://github.com/odoo/odoo/blob/saas-17.2/addons/web_editor/static/src/js/wysiwyg/wysiwyg_iframe.js#L82) set through JS so I just redefined the rule as `!important`.

I tested and it looks fine :slightly_smiling_face: 
What do you think about it?
Thanks!

tde-banana-odoo commented:
@rabodu r+

awa-odoo commented:
Hello @amdi-odoo 

Had a look at this one this morning.
At first it looked nice, the fixed width matches exactly what we need when creating a mailing.

However, it breaks easily, as adding a single field in the form would add a scrollbar (here I made "create_uid" visible just for testing purposes):
![image](https://github.com/user-attachments/assets/3d74b992-013d-4e11-9fef-3cad91ae4e47)

We could also have issues based on the height of the info/warning banner on top of the form (either a blank space or another scrollbar):
![image](https://github.com/user-attachments/assets/3f2805b1-7ef7-44fb-8819-31b70f1e029e)

It's still a better behavior that the current master, but I wonder if we could not make it work for all use cases while we're fixing this.
Did you have the same issues with the more "complex" solution (ResizeObserver) ?

Also, as this turns out to be much more complicated than initially thought, let's make another task for proper followup and more in-depth testing (I'll create a it).

Thanks & cheers!

amdi-odoo commented:
Hello @awa-odoo 
I don't have the issue when using the resize observer: https://tinyurl.com/2cohbccc
I think that basically, setting 70vh is too much of a fixed value, changing the iframe available space will just make it break.
What's great about the resize observer is that it will always compute the height based on the iframe content with a min height computed based on the iframe top offset/window height making it much more precised.

amdi-odoo commented:
@awa-odoo @tde-banana-odoo 
Created a new task here for the compressed mode: https://www.odoo.com/odoo/project/10888/tasks/4086474?cids=1 
As the fix is a "sparadrap", and we still need to discuss about changing all the `lg` bootstrap rules of the templates to `md` ones in stable :thinking: 
It's also better to split the 2 issues.
Cheers!

amdi-odoo commented:
Hello @awa-odoo @tde-banana-odoo @abd-msyukyu-odoo 

I gave another go on finding a css solution but as abd mentioned it is too much of a mess to "fight" against the form view and ancestors style. Also if we fix it with css, a single css change or a xml change could break the display => it is not a robust solution and is really messy.
(example: the iframe and website snippets are in different divs from 17.3 and above)

I kept the js approach as the iframe and website snippets height will be adapted no matter their xml placement.
Their height will also be dynamic, meaning adding a snippet on the template or adding a field in the form / adding an alert dialog won't break the display.

abd previously spotted an issue with my code:
"for big templates, one would have to scroll up to interact with the snippets, then scroll down again to drop a block at the bottom of the iframe"
Following his feedback, I changed the resizing computation to be sure both the iframe and website snippets are being resized.

What do you think?
Cheers!

awa-odoo commented:
Hello @amdi-odoo 

That seems fair, in the end it's a slightly more complex problem than initially thought so I don't mind having a few dozens of JS lines to try to fix it properly. It's important that it keeps looking nice and usable until we have a better / different solution with the phoenix editor.

I put the task in validation greened to try to speed up a bit as we have been working on this one for a good while.
As I will be off next week, @abd-msyukyu-odoo can do the review / r+ (once BE tested), if he doesn't mind.

Thanks for your thorough investigations & explanations on this one.

Cheers!

amdi-odoo commented:
Hello @abd-msyukyu-odoo  :slightly_smiling_face: 
cc @awa-odoo 

You were worried about the drag and drop of the snippets for big templates so I resized both the iframe and the website snippets to have 2 vertical scrollbars but apparently lna spotted that it breaks the [drag to scroll](https://www.awesomescreenshot.com/video/30101340?key=06de360bf5afc3ac152305a2913440fe ) behavior.

=> The drag to scroll behavior is pretty important so I should remove these 2 scrollbars to only have 1 on the website snippets and one on the full form.

I also spotted that, in 17.1, when we scrolled, the website snippets on the side was following the scrolling movement to ease the drag and drop (in 17.2 the website snippets doesn't follow the scrolling) : https://tinyurl.com/2aox65e9 

Because of this and all the mess we talked about in the above comments, I suggest we just bring back the deleted js part that was previously handling the resizing pretty perfectly. Typically the updateIframe, resize observer, the scroll event listeners and the `_resizeMailingEditorIframe` and `_repositionMailingEditorSidebar` methods.
[[1]](https://github.com/odoo/odoo/commit/cf6cd5ef46287f4e71fe96ad6b404ba9b08c1066#diff-2d45c4a0afb9cf14f222fbb26e9fbab6c736eb2d62d6eedd04bd4bc0393a9fe5L16-L31)   [[2]](https://github.com/odoo/odoo/commit/cf6cd5ef46287f4e71fe96ad6b404ba9b08c1066#diff-2d45c4a0afb9cf14f222fbb26e9fbab6c736eb2d62d6eedd04bd4bc0393a9fe5L36-L91)  [[3]](https://github.com/odoo/odoo/commit/cf6cd5ef46287f4e71fe96ad6b404ba9b08c1066#diff-2d45c4a0afb9cf14f222fbb26e9fbab6c736eb2d62d6eedd04bd4bc0393a9fe5L117-L152)

We'll then find a better/different solution in master for the new phoenix editor.
I think it'll be way easier and we won't have any broken feature or any future issue.

Cheers!

abd-msyukyu-odoo commented:
@amdi-odoo Hello, sorry for the delay, ok for reverting back to avoid the scroll and since Luc prefers one scrollbar :)

abd-msyukyu-odoo commented on addons/mass_mailing/static/src/js/mass_mailing_html_field.js:
```diff
@@ -34,6 +35,29 @@ export class MassMailingHtmlField extends HtmlField {
+        const throttledRepositionSidebar = useThrottleForAnimation(
+            this._repositionMailingEditorSidebar.bind(this)
+        );
+        onMounted(() => {
+            $('.o_content').on('scroll.repositionMailingEditorSidebar', throttledRepositionSidebar);
+            $('.o_form_sheet_bg').on('scroll.repositionMailingEditorSidebar', throttledRepositionSidebar);
+        });
+        onWillUnmount(() => {
+            $('.o_content').off('.repositionMailingEditorSidebar');
+            $('.o_form_sheet_bg').off('.repositionMailingEditorSidebar');
```
I believe here, the concept you went with is: for every known "scrollable", listen to the scroll event, then re-dimension, right ?

Can't we add  a listener to the scroll event on `document` only, and check if the `event.target` contains the iframe instead (to know when to execute the callback) ? (this will be more robust in case more scrollable are added, or if their classes change)

You can use `useExternalListener` which will handle `onMounted and onWillUnmount` and binding the function

abd-msyukyu-odoo reviewed (commented):
@amdi-odoo Heya! I commented with 2 ideas, let me know what you think :)

Also, in general I would be against adding more `jQuery` usage, but this file definitely has to be rewritten for the new editor anyway, so I'll let you decide if you want to take the time to convert your `jQuery` to `javascript`.

Cheers!

abd-msyukyu-odoo commented on addons/mass_mailing/static/src/js/mass_mailing_html_field.js:
```diff
@@ -34,6 +36,20 @@ export class MassMailingHtmlField extends HtmlField {
+        this.onIframeUpdated = () => {
+            onIframeUpdated();
+            this._updateIframe();
+        }
```
missing `;`

abd-msyukyu-odoo commented on addons/mass_mailing/static/src/js/mass_mailing_html_field.js:
```diff
@@ -168,6 +184,96 @@ export class MassMailingHtmlField extends HtmlField {
+        this.iframe = iframe;
+        this._resizeMailingEditorIframe();
+
+        const iframeTarget = this.iframe.contentDocument.querySelector('#iframe_target');
```
Not very important, but the JS lint configuration used in Odoo suggests always using `"` by default, unless unapplicable (this remark is valid for all new strings in this commit)

(you can enable the configuration for yourself using `odoo/addons/web/tooling/enable.sh` if you want)

abd-msyukyu-odoo commented on addons/mass_mailing/static/src/js/mass_mailing_html_field.js:
```diff
@@ -168,6 +184,96 @@ export class MassMailingHtmlField extends HtmlField {
+        const iframeTarget = this.iframe.contentDocument.querySelector('#iframe_target');
+        const elementToResize = themeSelectorNew || iframeTarget;
+        if (elementToResize) {
+            this.iframe.parentNode.style.height = `${this._isFullScreen() ? window.innerHeight : Math.max(elementToResize.scrollHeight, minHeight)}px`;
```
multiline ?

abd-msyukyu-odoo reviewed (commented):
@amdi-odoo Seems pretty good to me, I only have linting comments left :) Thanks!

abd-msyukyu-odoo commented:
@robodoo r+
